Libraries/rocSPARSE/preconditioner/csritilu0/main.cpp

```diff
@@ -0,0 +1,310 @@
+    constexpr std::array<rocsparse_int, n + 1> h_csr_row_ptr = {0, 6, 12, 18, 24, 30, 36};
+
+    // CSR column indices vector.
+    constexpr std::array<rocsparse_int, nnz> h_csr_col_ind
+        = {0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5,
+           0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5};
+
+    // 2. Allocate device memory and offload input data to the device.
+    rocsparse_int* d_csr_row_ptr{};
+    rocsparse_int* d_csr_col_ind{};
+    double*        d_csr_val{};
+    double*        d_LU{};
+    double*        d_data{};
+
+    constexpr size_t size_csr_row_ptr = sizeof(*d_csr_row_ptr) * (n + 1);
+    constexpr size_t size_csr_col_ind = sizeof(*d_csr_col_ind) * nnz;
+    constexpr size_t size_csr_val     = sizeof(*d_csr_val) * nnz;
+    constexpr size_t length_LU        = n * n;
+    constexpr size_t size_LU          = sizeof(*d_LU) * length_LU;
+    rocsparse_int    max_iter         = 100; /*maximum number of iterations*/
+    const size_t     length_data      = 2 * max_iter;
+    const size_t     size_data        = sizeof(*d_data) * length_data;
+
+    HIP_CHECK(hipMalloc(&d_csr_row_ptr, size_csr_row_ptr));
+    HIP_CHECK(hipMalloc(&d_csr_col_ind, size_csr_col_ind));
+    HIP_CHECK(hipMalloc(&d_csr_val, size_csr_val));
+    HIP_CHECK(hipMalloc(&d_LU, size_LU));
+    HIP_CHECK(hipMalloc(&d_data, size_data));
+
+    HIP_CHECK(
+        hipMemcpy(d_csr_row_ptr, h_csr_row_ptr.data(), size_csr_row_ptr, hipMemcpyHostToDevice));
+    HIP_CHECK(
+        hipMemcpy(d_csr_col_ind, h_csr_col_ind.data(), size_csr_col_ind, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_csr_val, h_csr_val.data(), size_csr_val, hipMemcpyHostToDevice));
+
+    // 3. Initialize rocSPARSE and prepare utility variables for csritilu0 invocation.
+    // Initialize rocSPARSE by creating a handle.
+    rocsparse_handle handle;
+    ROCSPARSE_CHECK(rocsparse_create_handle(&handle));
+
+    // Matrix descriptor.
+    rocsparse_mat_descr descr;
+    ROCSPARSE_CHECK(rocsparse_create_mat_descr(&descr));
+
+    // Iterative ILU0 (incomplete LU factorization with 0 fill-ins) algorithm used.
+    constexpr rocsparse_itilu0_alg alg = rocsparse_itilu0_alg_default;
+
+    // Options selected for the iterative algorithm:
+    //   rocsparse_itilu0_option_stopping_criteria      (2)
+    // + rocsparse_itilu0_option_compute_nrm_correction (+ 4)
+    // + rocsparse_itilu0_option_compute_nrm_residual   (+ 8)
+    // + rocsparse_itilu0_option_convergence_history    (+ 16)
+    // = 30
+    constexpr rocsparse_int option = 30;
+
+    // Tolerance.
+    constexpr double tol = 1.0e5 * std::numeric_limits<double>::epsilon(); /*2.22045e-11*/
+
+    // Indexing: zero based.
+    constexpr rocsparse_index_base idx_base = rocsparse_index_base_zero;
+
+    // Data type: 64 bit floating point real.
+    constexpr rocsparse_datatype data_type = rocsparse_datatype_f64_r;
+
+    // 4. Sort CSR matrix before calling any of the iterative-ILU0-related functions.
```

Comment:
why do you sort the matrix ? is it really needed ?

Reply:
We think that it is needed, because of the documentation:
https://rocsparse.readthedocs.io/en/master/usermanual.html#rocsparse-csritilu0-compute

Reply:
but the matrix you are building in the example is already sorted. I don't think this is needed.
I think csrilu0 also requires the matrix to be sorted and you don't sort.

Reply:
I think for sake of completeness of the example, having the sort explicit would make the requirement more clear. This allows people who try the examples to change values within the matrix without breaking the rest of the algorithm.